DESIGN.md

```diff
@@ -7,90 +7,64 @@
-    - `Call<Void> accept(List<Span> spans)`
-* `QueryRequest`: which includes
-    - `String serviceName, spanName;`
-    - `Map<String, String> annotationQuery;`
-    - `Long minDuration, maxDuration;`
-    - `long endTs, lookback;`
-    - `int limit;`
+* More focused on supporting processing than storage: traces and dependency links are emitted 
+downstream to support metrics aggregation. Storage is currently supported but in a single node.
 
 ### Kafka Zipkin Storage
 
-#### `KafkaSpanStore`
-
-Span Store is expecting Spans to be stored in topics partitioned by `TraceId`.
-
-> These can be created by Span Consumer, or can be **enriched** by other Stream Processors, outside of
-Zipkin Server.
-
-Kafka Span Store will need to support different kind of queries:
-
-
-##### Get Service Names/Get Span Names
-
-Service name to Span names pairs are indexed by aggregating spans.
-
-##### Get Trace/Find Traces
-
-When search requests are received, span index is used to search for trace ids. After a list is 
-retrieved, trace DAG is retrieved from trace state store.
+#### `KafkaSpanConsumer`
 
-##### Get Dependencies
+This component take batches of `spans` collected on different transports (e.g. HTTP, Kafka) .
 
-After `spans` are aggregated into traces, traces are processed to collect dependencies. 
-Dependencies changelog are stored in a Kafka topic to be be stored as materialized view on 
-Zipkin instances.
+In order to support aggregation and post-processing, batches are `flatmap`ing into individual 
+spans, keyed by `trace ID`.
 
-### Stream processors
+#### Stream processors
 
 #### Trace Aggregation Stream Processor
 
-This is the main processors that take incoming spans and aggregate them into:
-
+This processor take incoming partitioned spans and aggregate them into:
 - Traces
 - Dependencies
 
-![service aggregation](docs/service-aggregation-stream.png)
+**Trace completion:** A key part of the stream processing pipeline is define when a trace is completed
+to emit its complete state. This is done by using [Session windows](https://kafka.apache.org/23/javadoc/org/apache/kafka/streams/kstream/SessionWindows.html)
+grouping spans by trace ID, and waiting for a certain inactivity period without receiving messages 
+to emit a trace.
 
-![dependency aggregation](docs/dependency-aggregation-stream.png)
+> One **main consideration** with this approach is that in order to evaluate inactivity gap, a new 
+> event has to be received. This means if you send a set of trace spans and wait for it to be shown 
+> as query result, it requires another span passed the inactivity gap to be trigger.
+> 
+> This might not be what you expect when using Zipkin for demo purposes, but is based on the assumption
+> that your applications are continuously emitting spans, so aggregation happens continuously.
+> For more info: <https://stackoverflow.com/questions/54222594/kafka-stream-suppress-session-windowed-aggregation/54226977#54226977> 
```

Comment:
how do people normally accommodate this? would a synthetic span heartbeat be a bad idea (that could be disabled)

Reply:
The SO comment mentions that as the stream keeps moving, the windows will get closed eventually, so in a production context you will get "completed" traces constantly.

Another option could be to remove `suppress` operator conditionally (https://github.com/jeqo/zipkin-storage-kafka/pull/25/files#diff-e1fbe07c15a6efece3ac2119c54eefa7R85) 
But this will lead every change on the window to be emitted downstream, leading to incorrect results in dependencies stream (e.g. counting same links many times); as well as not helping on reducing the load downstream.

Also windowing timeouts are configurable. I reduce sizes on docker composes to make testing easier.

The idea of a heartbeat sounds interesting; but right now I don't see where it could be plugged: (thinking out loud) we could add another KStraem app consuming from spans and schedule emitting this syntactic span every inactivity gap + 1, and only emit these to the same topic to kick out the supressing in the another streams. Anyhow, will keep this in mind.

Would make sense to position this storage as "eventual consistent" and focus on enable processing, instead of supporting "dev" use-cases like post a trace and query immediately, like mem db?

Reply:
I would recommend moving your last comment to a new issue for now, as it keeps the meaty discussions partitioned. Most important I think is merging the work here.. kinks can come out later.